File: public/dsa/linked_lists/remove_duplicates_from_sorted_linked_list/approach/remove_duplicates_from_sorted_linked_list.cpp

#include <bits/stdc++.h>
using namespace std;

// defs-start
class ListNode {
   public:
    int data;
    ListNode* next;
    ListNode(int data, ListNode* next = nullptr) : data(data), next(next) {}
};

// defs-end

ListNode* removeDuplicates(ListNode* head) {
    // add your logic here
    auto temp = head;
    while (temp && temp->next) {
        if (temp->data == temp->next->data) {
            temp->next = temp->next->next;
        } else {
            temp = temp->next;
        }
    }
    return head;
}

// tests-start

ListNode* formList(const vector<int>& arr) {
    if (arr.empty()) return nullptr;

    ListNode* head = new ListNode(arr[0]);
    ListNode* curr = head;

    for (int i = 1; i < arr.size(); ++i) {
        curr->next = new ListNode(arr[i]);
        curr = curr->next;
    }

    return head;
}

void printList(ListNode* head) {
    ListNode* temp = head;
    while (temp) {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;
}

int main() {
    ListNode* head = formList({1, 2, 2, 2, 3, 3, 4, 7});

    ListNode* res = removeDuplicates(head);

    printList(res);

    return 0;
}
// tests-end

/*output
1 2 3 4 7
*/

File: public/dsa/linked_lists/remove_duplicates_from_sorted_linked_list/approach2/remove_duplicates_from_sorted_linked_list.py
from typing import Optional, List

# defs-start


class ListNode:
    def __init__(self, data=0, next=None):
        self.data = data
        self.next = next
# defs-end


class Solution:
    def removeDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current = head
        while current:
            temp = current.next
            while temp and temp.data == current.data:
                temp = temp.next
            current.next = temp
            current = temp
        return head


# tests-start
def formList(arr: List[int]):
    if len(arr) == 0:
        return None

    head = ListNode(arr[0])
    curr = head
    for i in range(1, len(arr)):
        curr.next = ListNode(arr[i])
        curr = curr.next
    return head


def printList(head: Optional[ListNode]):
    temp = head
    while temp:
        print(temp.data, end=' ')
        temp = temp.next
    print()


head = formList([1, 2, 2, 2, 3, 3, 4, 7])
sol = Solution()
res = sol.removeDuplicates(head)
printList(res)
# tests-end

'''output
1 2 3 4 7
'''
